### PR TITLE
Automated framing in Steel

### PR DESCRIPTION
**Pull request by R1kM**
This PR merges automated framing inside the steel branch. Some parts are still WIP, but this at least unlocks work in parallel on libraries with the new effect.

Several files previously verifying with manual framing were not ported yet, and were moved to examples/steel/effects. These files are Steel.Channel.Simplex, Steel.Closure and Steel.MonotonicHigherReference. The old effects are also in examples/steel/effects

Path forward before a full merge to master:
* Revive the old Steel effects with manual framing as a SteelX effect, with a polymonadic subcomp from SteelX to Steel following Aseem's proposal. This would enable manual framing on demand.
* Reexpose files moved to examples/steel/effects by keeping the implementation as a SteelX, and performing a subcomp to give a Steel interface
* Implement Steel.Effect and Steel.Effect.Atomic. Almost all of the missing parts are direct reflections from actions to the Steel(Atomic) effect(s). The easiest would also be to expose them as SteelX functions, and call an explicit lift.
* Finish implementing Steel.Duplex.PCM to fit the interface of Steel.Channel.Duplex
* Remove the last assumes in Steel.Duplex.PCM/Steel.Stepper due to universe difficulties with slprops/pts_to, and to implementations of actions using fictional separation logic
* Implement `sl_implies` and the corresponding lemmas in Steel.Effect.Common
* Restore examples in steel/examples for which verification was removed

### CODE REVIEW

**Comment by R1kM** [comment]
Almost all items from the previous TODO list are done.
The only remaining part is to wrap up Steel.Duplex.PCM. It currently has two assumes for updates relying on fictional separation logic that need to be implemented. It also doesn't match completely the Steel.Duplex interface yet. I moved this file to "examples/steel" for now, so this is non-blocking.

During today's Steel meeting, we agreed to proceed as follows to merge this PR into steel, and then master: Some unifier changes which are also partially on this branch (such as https://github.com/FStarLang/FStar/pull/2152) need to land first in master. We will then merge master into steel, and solve conflicts on this branch before propagating upstream.

**Comment by nikswamy** [comment]
Thanks Aymeric! This is a great milestone ... congrats! I'm merging it now.

File: src/tactics/FStar.Tactics.Interpreter.fs [hunk 1]
```diff
@@ -504,8 +510,13 @@ let run_tactic_on_ps
     Err.stop_if_err ();
```
Comment (nikswamy) [inline src/tactics/FStar.Tactics.Interpreter.fs:510]: FYI, this chunk of changes to run_tactic_on_ps will not survive to master. We will have a separate PR into master that revises this and then incorporate those changes here.

File: src/typechecker/FStar.TypeChecker.DeferredImplicits.fs [hunk 2]
```diff
@@ -138,7 +138,7 @@ let solve_deferred_to_tactic_goals env g =
         in
         let goal_ty = U.mk_eq2 (env.universe_of env_lax t_eq) t_eq tp.lhs tp.rhs in
         let goal, ctx_uvar, _ =
-            Env.new_implicit_var_aux reason tp.lhs.pos env goal_ty Strict None
+            Env.new_implicit_var_aux reason tp.lhs.pos env goal_ty Allow_untyped None
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.DeferredImplicits.fs:141]: This is an important one. @aseemr @mtzguido The rationale is that this goal is being solved by a tactic and so doesn't need to be re-checked.

File: src/typechecker/FStar.TypeChecker.Rel.fs [hunk 3]
```diff
@@ -4063,36 +4063,42 @@ let resolve_implicits' env is_tac g =
...
-                               (Print.term_to_string ctx_u.ctx_uvar_typ)
-                               reason
-                               (Range.string_of_range r);
-               let g =
-                 try
-                   env.check_type_of must_total env tm ctx_u.ctx_uvar_typ
-                 with e when Errors.handleable e ->
-                    Errors.add_errors [Error_BadImplicit,
-                                       BU.format3 "Failed while checking implicit %s set to %s of expected type %s"
+               if forcelax
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Rel.fs:4066]: FYI, this change will not survive to master. We will revise it separately in master and merge here.

**Pull request merged**